### PR TITLE
Copter: run rate loop at full filter rate in its own thread

### PR DESCRIPTION
**Pull request by tridge**
This adds FLIGHT_OPTIONS=8 which creates a rate loop thread, allowing rate updates at the full filtered gyro rate (2kHz on most copters), settable with existing INS_GYRO_RATE parameter.
This was test flown successfully on a 5" quad from @MichelleRos, the rate loop did run at 2kHz (I added a send text for average rate every 2s)

We've now had successful flights at 2kHz, 4kHz and 8kHz on a MatekH743-bdshot 5" quad. At 8kHz we had to disable the 2nd IMU with IMU_ENABLE_MASK or we had R/C issues

### CODE REVIEW

**Comment by andyp1per** [comment]
I had a go at flying this but nearly had a fly away due to frame resonance. My guess would be that the PID notch sample frequency is not being updated correctly with the right values but have not looked in detail at what is going on.

**Comment by tridge** [comment]
> I had a go at flying this but nearly had a fly away due to frame resonance. My guess would be that the PID notch sample frequency is not being updated correctly with the right values

you're right, it was using the wrong frequency for PID notch, I've fixed it in the PR

File: ArduCopter/Attitude.cpp [hunk 1]
```diff
@@ -4,18 +4,100 @@
...
+    uint32_t last_run_us = AP_HAL::micros();
+    float dt_avg = 0.0;
+    uint32_t now_ms = AP_HAL::millis();
+    uint32_t last_report_ms = now_ms;
+    uint32_t last_notch_sample_ms = now_ms;
+    bool was_using_rate_thread = false;
+
+    while (true) {
+        // allow changing option at runtime
+        if (!flight_option_is_set(FlightOptions::USE_RATE_LOOP_THREAD)) {
```
Comment (andyp1per) [inline ArduCopter/Attitude.cpp:24]: I think in the final PR you should make this a boot time option only. It increases the complexity of the code and I'm not sure it really adds a lot of value.

File: ArduCopter/Copter.cpp [hunk 2]
```diff
@@ -674,11 +674,22 @@ void Copter::one_hz_loop()
...
+    }
     pos_control->get_accel_z_pid().set_notch_sample_rate(AP::scheduler().get_filtered_loop_rate_hz());
 #if AC_CUSTOMCONTROL_MULTI_ENABLED == ENABLED
     custom_control.set_notch_sample_rate(AP::scheduler().get_filtered_loop_rate_hz());
 #endif
+
+    // see if we should have a separate rate thread
+    if (!started_rate_thread && flight_option_is_set(FlightOptions::USE_RATE_LOOP_THREAD)) {
+        if (hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&Copter::rate_controller_thread, void),
+                                         "rate_controller",
```
Comment (andyp1per) [inline ArduCopter/Copter.cpp:688]: Please can we shorten the thread name as it gets truncated in ```@SYS/threads.txt```

**Review by andyp1per — CHANGES_REQUESTED** [review]
Looks good. I think the main thing for me is that we should call AP_Vehicle::update_dynamic_notch_at_specified_rate() from the rate loop thread since these are all filtering related. Also without this people will not be able to significantly lower their loop rate without losing update rates. We obviously would need some options to control the actual rate.

**Comment by tridge** [comment]
> I think the main thing for me is that we should call AP_Vehicle::update_dynamic_notch_at_specified_rate() from the rate loop thread since these are all filtering related

easy to do, we just need to assess the CPU cost to see how often we can call it reasonably

**Comment by MichelleRos** [comment]
> we just need to assess the CPU cost to see how often we can call it reasonably

I'd actually be curious to test how much this impacts performance - maybe updating the notches at the rate loop rate gives better performance than leaving the notches updating slowly and running the PIDs faster.

**Comment by andyp1per** [comment]
Quick test - now seems to fly fine. Couple of issues:
- Log download seems to be twice as slow
- Dshot tone commands seem to be missing, still get tones but some are missing

File: ArduCopter/Attitude.cpp [hunk 3]
```diff
@@ -4,18 +4,113 @@
...
+
+        // wait for an IMU sample
+        rate_sem.wait_blocking();
+        if (ap.motor_test) {
+            continue;
+        }
+
+        const uint32_t now_us = AP_HAL::micros();
+        const uint32_t dt_us = now_us - last_run_us;
+        const float dt = dt_us * 1.0e-6;
```
Comment (amilcarlucas) [inline ArduCopter/Attitude.cpp:48]:
```suggestion
        const float dt = dt_us * 1.0e-6f;
```

**Comment by andyp1per** [comment]
I tried running autotune on my copter with this change and it completely failed. Very strange noises coming from the motors on twitches. Autotune picked minimum gains for everything. So there is definitely something not right in the implementation that needs to be resolved before I can dig in to whether it makes a worthwhile difference or not.

**Comment by andyp1per** [comment]
So the problem appeared to be extreme jitter in the motor outputs cause by extreme jitter in the SPI reads which never actually got anywhere the required rate (running at about 1.6Khz vs 4Khz), so I thin that made all the maths be messed up. This was ultimately caused by the SPI thread not having a high enough priority and the device scheduler introducing arbitrary delays. I have made some changes to resolve these and the branch is here - https://github.com/andyp1per/ardupilot/tree/pr-fast-rate-thread (includes a rebase). The output now looks pretty regular at 4Khz. Too wet to fly, but indoor testing looks promising.

**Comment by andyp1per** [comment]
Did a quick flight test outside - certainly the grinding noises are all but gone. Will need to see if I can successfully do an autotune but the log looks promising

**Comment by tridge** [comment]
@andyp1per interesting! hope a full flight test goes well

**Comment by andyp1per** [comment]
Still wouldn't tune. Its certainly a lot better, but still something strange going on. I also tried using the actual dt rather than the average and had a flyaway - would barely respond to RC input. There is a real danger of thread starvation of the main loop with the higher priority threads which is what I just saw 😮

**Comment by andyp1per** [comment]
@lthall has pointed out that autotune is not zeroing D_FF (which I will fix) and the target is showing spurious changes:

image

**Comment by tridge** [comment]
replaced by #27029

**Pull request closed without merging**